src/reconcile/rec_visibility.c

```diff
@@ -586,35 +587,70 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
         upd_select->start_ts = WT_TS_NONE;
         upd_select->start_txn = WT_TXN_NONE;
         upd_select->stop_ts = WT_TS_MAX;
         upd_select->stop_txn = WT_TXN_MAX;
-        if (upd_select->upd->start_ts != WT_TS_NONE)
+        /*
+         * If the newest is a tombstone then select the update before it and set the end of the
+         * visibility window to its time pair as appropriate to indicate that we should return "not
+         * found" for reads after this point.
+         *
+         * Otherwise, leave the end of the visibility window at the maximum possible value to
+         * indicate that the value is visible to any timestamp/transaction id ahead of it.
+         */
+        WT_ASSERT(session, upd_select->upd != NULL);
+        if (upd_select->upd->type == WT_UPDATE_TOMBSTONE) {
+            upd_select->stop_ts = upd_select->upd->start_ts;
+            upd_select->stop_txn = upd_select->upd->txnid;
+            upd_select->upd = upd_select->upd->next;
+        }
+        if (upd_select->upd != NULL) {
+            /* The beginning of the validity window is the selected update's time pair. */
             upd_select->durable_ts = upd_select->start_ts = upd_select->upd->start_ts;
-        if (upd_select->upd->txnid != WT_TXN_NONE)
             upd_select->start_txn = upd_select->upd->txnid;
+        } else {
+            /*
+             * The only time we don't get the unpacked value is when we insert a key in which case
+             * we can't have a tombstone.
+             */
+            WT_ASSERT(session, vpack != NULL);
+            /*
+             * It's possible to have a tombstone as the only update in the update list. If we
+             * reconciled before with only a single update and then read the page back into cache,
+             * we'll have an empty update list. And applying a delete on top of that will result in
+             * ONLY a tombstone in the update list.
+             *
+             * In this case, we should leave the selected update unset to indicate that we want to
+             * keep the same on-disk value but set the stop time pair to indicate that the validity
+             * window ends when this tombstone started.
+             */
+            upd_select->durable_ts = upd_select->start_ts = vpack->start_ts;
+            upd_select->start_txn = vpack->start_txn;
+            /*
+             * Leaving the update unset means that we can skip reconciling. If we've set the stop
+             * time pair because of a tombstone after the on-disk value, we still have work to do so
+             * that is NOT ok. Let's allocate an update equivalent to the on-disk value and continue
+             * on our way!
+             */
+            WT_RET(__wt_scr_alloc(session, 0, &tmp));
+            WT_ERR(__wt_page_cell_data_ref(session, page, vpack, tmp));
+            WT_ERR(__wt_update_alloc(session, tmp, &upd_select->upd, &size, WT_UPDATE_STANDARD));
+            upd_select->upd->ext = 1;
```

Comment:
Do all the WT_UPDATE are allocated with external memory contents? Do we need this special flag?

Reply:
The `ext` flag is to signal to callers of `__wt_rec_upd_select` that they are responsible for freeing the update. At the moment, this can happen if reconciliation chooses a lookaside update (rather than one from the in-memory update list which you would NOT want to free). In this case, we're allocating an update specifically for reconciliation purposes so we'll need to free it when we're done with it.

Reply:
We need to get rid of `WT_UPDATE.ext`, it's not a good thing. I think @sulabhM has a plan...

Reply:
Yeah, I am using `ext` because it fits in the current view of the world. But it does need to be addressed at some point.